File: java/circular-buffer/src/main/java/org/mydotey/util/TimeSequenceCircularBufferConfig.java
package org.mydotey.util;

/**
 * @author koqizhao
 *
 * Jul 25, 2018
 */
public class TimeSequenceCircularBufferConfig implements Cloneable {

    private long _timeWindow;
    private long _bucketTtl;
    private int _bucketCount;

    protected TimeSequenceCircularBufferConfig() {

    }

    public long getTimeWindow() {
        return _timeWindow;
    }

    public long getBucketTtl() {
        return _bucketTtl;
    }

    public int getBucketCount() {
        return _bucketCount;
    }

    @Override
    public TimeSequenceCircularBufferConfig clone() {
        try {
            return (TimeSequenceCircularBufferConfig) super.clone();
        } catch (CloneNotSupportedException e) {
            e.printStackTrace();
            return null;
        }
    }

    @Override
    public String toString() {
        return String.format("%s { timeWindow: %s, bucketTtl: %s, bucketCount: %s }", getClass().getSimpleName(),
                _timeWindow, _bucketTtl, _bucketCount);
    }

    public static class Builder extends AbstractBuilder<Builder> {

    }

    @SuppressWarnings("unchecked")
    public static abstract class AbstractBuilder<B extends AbstractBuilder<B>> {

        private TimeSequenceCircularBufferConfig _config;

        protected AbstractBuilder() {
            _config = newConfig();
        }

        protected TimeSequenceCircularBufferConfig newConfig() {
            return new TimeSequenceCircularBufferConfig();
        }

        protected TimeSequenceCircularBufferConfig getConfig() {
            return _config;
        }

        public B setTimeWindow(long timeWindow) {
            _config._timeWindow = timeWindow;
            return (B) this;
        }

        public B setBucketTtl(long bucketTtl) {
            _config._bucketTtl = bucketTtl;
            return (B) this;
        }

        public TimeSequenceCircularBufferConfig build() {
            if (_config._timeWindow <= 0)
                throw new IllegalArgumentException("timeWindow not set");

            if (_config._bucketTtl <= 0)
                throw new IllegalArgumentException("bucketTtl not set");

            if (_config._timeWindow % _config._bucketTtl != 0)
                throw new IllegalArgumentException(String.format("timeWindow %s cannot be divided by bucketTtl %s.",
                        _config._timeWindow, _config._bucketTtl));

            long bucketCount = _config._timeWindow / _config._bucketTtl;
            if (bucketCount > Integer.MAX_VALUE)
                throw new IllegalArgumentException(String.format("timeWindow %s too large", _config._timeWindow));

            _config._bucketCount = (int) bucketCount;
            return _config.clone();
        }

    }

}

File: dotnet/src/MyDotey.CircularBuffer/TimeSequenceCircularBufferConfig.cs
using System;

namespace MyDotey.Util
{
    /**
     * @author koqizhao
     *
     * Jul 25, 2018
     */
    public class TimeSequenceCircularBufferConfig : ICloneable
    {
        private long _timeWindow;
        private long _bucketTtl;
        private int _bucketCount;

        protected TimeSequenceCircularBufferConfig()
        {

        }

        public long TimeWindow
        {
            get { return _timeWindow; }
        }

        public long BucketTtl
        {
            get { return _bucketTtl; }
        }

        public int BucketCount
        {
            get { return _bucketCount; }
        }

        public object Clone()
        {
            return MemberwiseClone();
        }

        public override String ToString()
        {
            return String.Format("{0} {{ timeWindow: {1}, bucketTtl: {2}, bucketCount: {3} }}", GetType().Name,
                    _timeWindow, _bucketTtl, _bucketCount);
        }

        public class Builder : AbstractBuilder<Builder, TimeSequenceCircularBufferConfig>
        {

        }

        public abstract class AbstractBuilder<B, C>
            where B : AbstractBuilder<B, C>
        {
            private TimeSequenceCircularBufferConfig _config;

            protected AbstractBuilder()
            {
                _config = NewConfig();
            }

            protected virtual TimeSequenceCircularBufferConfig NewConfig()
            {
                return new TimeSequenceCircularBufferConfig();
            }

            protected TimeSequenceCircularBufferConfig Config
            {
                get { return _config; }
            }

            public B SetTimeWindow(long timeWindow)
            {
                _config._timeWindow = timeWindow;
                return (B)this;
            }

            public B SetBucketTtl(long bucketTtl)
            {
                _config._bucketTtl = bucketTtl;
                return (B)this;
            }

            public virtual C Build()
            {
                if (_config._timeWindow <= 0)
                    throw new ArgumentNullException("timeWindow not set");

                if (_config._bucketTtl <= 0)
                    throw new ArgumentNullException("bucketTtl not set");

                if (_config._timeWindow % _config._bucketTtl != 0)
                    throw new ArgumentException(String.Format("timeWindow {0} cannot be divided by bucketTtl {1}.",
                            _config._timeWindow, _config._bucketTtl));

                long bucketCount = _config._timeWindow / _config._bucketTtl;
                if (bucketCount > int.MaxValue)
                    throw new ArgumentException(String.Format("timeWindow {0} too large", _config._timeWindow));

                _config._bucketCount = (int)bucketCount;
                return (C)_config.Clone();
            }
        }
    }
}